lib/index.js

```diff
@@ -104,45 +117,78 @@ exports.parse = function (definition, options) {
+                        errors.push(internals.formatError('Invalid value (non-number) for option:', last.def.name));
                         continue;
                     }
                 }
 
-                if (last.valid &&
-                    last.valid.indexOf(value) === -1) {
+                if (last.def.valid &&
+                    !last.def.valid.includes(value)) {
 
                     const validValues = [];
-                    for (let j = 0; j < last.valid.length; ++j) {
-                        const valid = last.valid[j];
+                    for (let j = 0; j < last.def.valid.length; ++j) {
+                        const valid = last.def.valid[j];
                         validValues.push(`'${valid}'`);
                     }
 
-                    errors.push(internals.formatError('Invalid value for option:', last.name, '(valid: ' + validValues.join(',') + ')'));
+                    errors.push(internals.formatError('Invalid value for option:', last.def.name, '(valid: ' + validValues.join(',') + ')'));
                     continue;
                 }
+
+                if (last.def.type === 'json') {
+
+                    const stringValue = value;
+
+                    try {
+                        value = Bourne.parse(value, {   // E.g. { x }
+                            protoAction: 'remove'
+                        });
+                    }
+                    catch (err) {
+                        // If the input doesn't look like JSON, we'll ignore that and treat it as a string
+                        Bounce.ignore(err, SyntaxError);
+                    }
+
+                    if (!last.def.parsePrimitives && (!value || typeof value !== 'object')) {
+                        // When receiving JSON that parses to a non-object, treat it as a string, i.e. numbers, true, false, null.
+                        value = stringValue;
+                    }
+
+                    value = internals.valueAtArgPath(last.opt, value);
+
+                    if (!value || typeof value !== 'object') {
+                        errors.push(internals.formatError('Invalid value for option:', last.def.name, '(must be an object or array)'));
+                        continue;
+                    }
+
+                    // Necessary to protect from prototype poisoning in dot-separated path e.g. `--x.__proto__.y 1`
+                    Bourne.scan(value, { protoAction: 'remove' });
+                }
             }
 
-            const name = last ? last.name : '_';
+            const name = last ? last.def.name : '_';
             if (flags.hasOwnProperty(name)) {
 
                 if (!last ||
-                    last.multiple) {
+                    last.def.multiple) {
 
                     flags[name].push(value);
                 }
+                else if (last.def.type === 'json') {
+                    Hoek.merge(flags[name], value);
```

Comment:
If a default value was provided, the merge will modify it. I suspect the default value will need to be cloned to avoid surprises.

Reply:
Each definition is cloned near the top of `Bossy.parse()`, and I do have a test confirming that defaults aren't mutated, so I think we're good 👍 https://github.com/hapijs/bossy/pull/77/files#diff-5bb8db779819ddef5956a5d9d5949c05ef7445237656ca37bf2f02720271440bR362